src/AndersonAcceleration.f90

```diff
@@ -72,183 +71,288 @@ MODULE AndersonAccelerationTypes
+          tmpA=BLAS_dot(solver%tmpvec,x_new)
+          CALL solver%LS%A%set(k,i,tmpA)
+        ENDDO
+      ENDDO
+      CALL solver%LS%solve()
+      DO i=1,depth_s
+        CALL solver%LS%x%get(i,solver%alpha(i))
+      ENDDO
+    ENDIF
+
+    !Back out "0th" alpha
+    solver%alpha(depth_s+1)=1.0_SRK
+    DO i=1,depth_s
+      solver%alpha(depth_s+1)=solver%alpha(depth_s+1)-solver%alpha(i)
+    ENDDO
+
+    !Ensure Anderson coefficient solve did not fail due to bad input vector
+    IF(ISNAN(solver%alpha(depth_s+1)) .OR. (solver%alpha(1) .APPROXEQ. 0.0_SRK)) THEN
+      !Bad Anderson coefficient solve, revert to under-relaxation and reset Anderson solver
+      CALL x_new%set(0.0_SRK)
+      CALL BLAS_axpy(solver%Gx(1),x_new,solver%beta)
+      CALL BLAS_axpy(solver%x(1),x_new,1.0_SRK-solver%beta)
+      CALL solver%reset(x_new)
+    ELSE
+      !Get accelerated solution
+      CALL x_new%set(0.0_SRK)
+      DO i=1,depth_s+1
+        CALL BLAS_copy(solver%x(i),solver%tmpvec)
+        CALL BLAS_axpy(solver%r(i),solver%tmpvec,solver%beta)
+        CALL BLAS_axpy(solver%tmpvec,x_new,solver%alpha(i))
+      ENDDO
+
+      !Push back solution vectors and load in newest accelerated as next initial iterate
+      DO i=MIN(depth_s+1,solver%depth),1,-1
+        CALL BLAS_copy(solver%x(i),solver%x(i+1))
+      ENDDO
+      CALL BLAS_copy(x_new,solver%x(1))
+    ENDIF
   ELSE
-    CALL Anderson_Update(solver%id)
+    !Get under-relaxed solution using mixing parameter
+    CALL BLAS_scal(x_new,solver%beta)
+    CALL BLAS_axpy(solver%x(1),x_new,1.0_SRK-solver%beta)
+    CALL BLAS_copy(x_new,solver%x(1))
   ENDIF
 
-  solver%iter=solver%iter+1
-#endif
 ENDSUBROUTINE step_AndersonAccelerationType
 !
 !-------------------------------------------------------------------------------
-!> @brief Reset the Anderson acceleration solver
-!> @param solver The anderson solver to act on
-!>
-!> This routine resets the counters for the anderson acceleration
+!> @brief Set or reset the initial iterate and linear solver state for the Anderson solver
+!> @param solver The Anderson solver to act on
+!> @param x Initial iterate solve is starting from
 !>
-SUBROUTINE reset_AndersonAccelerationType(solver)
+SUBROUTINE reset_AndersonAccelerationType(solver,x)
+  CHARACTER(LEN=*),PARAMETER :: myName='reset_AndersonAccelerationType'
   CLASS(AndersonAccelerationType),INTENT(INOUT) :: solver
-#ifdef FUTILITY_HAVE_Trilinos_NOX
-  CALL Anderson_Reset(solver%id)
-  solver%iter=1
+  CLASS(VectorType),INTENT(INOUT) :: x
+
+  INTEGER(SIK) :: i,j
+
+  REQUIRE(x%n == solver%n)
+
+  !If this is the first call to set/reset must actually create vectors of needed type
+  IF(solver%s == 0) THEN
+    IF(.NOT.ALLOCATED(solver%x)) THEN
+
+      !!!TODO Once missing BLAS interfaces have been added for the following vector types
+      !do away with this error catch to allow use of these with Anderson Acceleration.
+#ifdef HAVE_MPI
+      SELECT TYPE(x);TYPE IS(NativeDistributedVectorType)
```

Comment:
`TYPE IS(...)` should go on its own line

Reply:
I agree. I'll change it here. We have a LOT of instances in MPACT with the ; for this.

Reply:
So the standard allows the `;` if there is only one branch.  But since you have 2 (or 3 with my proposed change to this section), each should be on its own line for consistency within the block.

Reply:
So I'm fine with what you did on this... if it works.  If we were to build without trilinos and without mpi, though, it would compile to this:
```
SELECT TYPE(x)
ENDSELECT
```
Can you have a SELECT block with no branches?  I'm not even sure.  I'm fine leaving it this way as long as it works for a non-mpi, non-trilinos build.